command-line-parser.dylan

```diff
@@ -167,90 +194,105 @@ define function pass-by-name-options
          cmd.command-options)
 end function;
 
-define open abstract class <subcommand> (<command>)
-  constant slot subcommand-name :: <string>,
-    required-init-keyword: name:;
-end class;
-
-define method debug-name
-    (subcmd :: <subcommand>) => (name :: <string>)
-  subcmd.subcommand-name
-end method;
-
-// Should this just be another method on execute-command instead?
-define open generic execute-subcommand
-    (parser :: <command-line-parser>, subcmd :: <object>)
- => (status :: false-or(<integer>));
-
-define method execute-subcommand
-    (parser :: <command-line-parser>, subcmd :: <object>)
- => (status :: false-or(<integer>))
-  error("don't know how to execute subcommand %=. add an execute-subcommand method?",
-        subcmd);
-end method;
-
-define open class <command-line-parser> (<command>)
-  slot parser-subcommands :: <sequence> = #[],
-    init-keyword: subcommands:;
-  slot selected-subcommand :: false-or(<subcommand>) = #f;
-end class;
-
-// Help options are on by default but may be turned off.
-define method initialize
-    (parser :: <command-line-parser>,
-     #key help-option? :: <boolean> = #t,
-          help-subcommand? :: <boolean> = #t, #all-keys) => ()
-  next-method();
-  if (help-option?)
-    add-help-option(parser);
-  end;
-  if (help-subcommand? & parser.has-subcommands?)
-    add-help-subcommand(parser);
-  end;
-end method;
-
 define function has-subcommands?
-    (parser :: <command-line-parser>) => (_ :: <boolean>)
-  parser.parser-subcommands.size > 0
+    (cmd :: <command>) => (_ :: <boolean>)
+  cmd.command-subcommands.size > 0
 end;
 
 define generic find-subcommand
-    (parser :: <command-line-parser>, object)
+    (cmd :: <command>, object)
  => (subcommand :: false-or(<subcommand>));
 
 define method find-subcommand
-    (parser :: <command-line-parser>, class :: subclass(<subcommand>))
+    (cmd :: <command>, class :: subclass(<subcommand>))
  => (subcommand :: false-or(<subcommand>))
-  let subs = parser.parser-subcommands;
+  let subs = cmd.command-subcommands;
   let key = find-key(subs, rcurry(instance?, class));
   key & subs[key]
 end method;
 
 define method find-subcommand
-    (parser :: <command-line-parser>, name :: <string>)
+    (cmd :: <command>, name :: <string>)
  => (subcommand :: false-or(<subcommand>))
-  let subs = parser.parser-subcommands;
+  let subs = cmd.command-subcommands;
   let key = find-key(subs, method (subcmd)
                              name = subcmd.subcommand-name
                            end);
   key & subs[key]
 end method;
 
+define method find-subcommand
+    (cmd :: <command>, name :: <sequence>)
+ => (subcommand :: false-or(<subcommand>))
+  iterate loop (cmd = cmd, names = as(<list>, name))
```

Comment:
Why convert to a list just to iterate through it?

Reply:
You make a fine point.  I guess I had other ideas at the time but I don't remember what they were so I'll fix this.

Reply:
Done